File: extensions/drm-copilot/resources/scripts/dev_tools/new_potential_bug_entry.py
"""Create a potential bug entry and optionally open it in VS Code."""

from __future__ import annotations

import argparse
import os
import re
import shutil
import subprocess
from dataclasses import dataclass
from datetime import date
from pathlib import Path
from typing import TYPE_CHECKING, Protocol

if TYPE_CHECKING:
    from collections.abc import Callable, Iterable

SHORT_NAME_PATTERN = re.compile(r"^[a-z0-9]+(-[a-z0-9]+)*$")
_INSIDERS_SIGNAL_NAMES = (
    "TERM_PROGRAM_VERSION",
    "VSCODE_GIT_ASKPASS_MAIN",
    "TERM_PROGRAM",
    "VSCODE_IPC_HOOK_CLI",
)


def _resolve_workspace() -> Path:
    return Path.cwd()


def validate_short_name(short_name: str) -> None:
    """Validate that a short name matches the repository kebab-case contract.

    Purpose:
        Reject invalid bug-entry short names before filesystem or template work begins.

    Args:
        short_name (str): Candidate short name that must use lowercase
            kebab-case tokens.

    Returns:
        None: This validator returns no value when the short name satisfies
            the contract.

    Raises:
        ValueError: Raised when the supplied short name is blank or
            violates the kebab-case pattern.

    Side Effects:
        None.
    """
    if not short_name or not SHORT_NAME_PATTERN.fullmatch(short_name):
        raise ValueError(
            f"Aborted: '{short_name}' is invalid. Use kebab-case letters/numbers only "
            "(e.g., api-timeout)."
        )


def default_git_config_lookup(key: str) -> str | None:
    """Resolve a Git configuration value without failing when Git is unavailable.

    Purpose:
        Query Git configuration opportunistically so author discovery can
            prefer repository settings.

    Args:
        key (str): Git configuration key to read, such as `user.name`.

    Returns:
        str | None: The trimmed configuration value when Git is available
            and returns a non-blank value; otherwise `None`.

    Raises:
        None.

    Side Effects:
        Invokes the `git config` subprocess when Git is available on PATH.
    """
    git_cmd = shutil.which("git")
    if not git_cmd:
        return None
    result = subprocess.run(  # noqa: S603
        [git_cmd, "config", key],
        check=False,
        capture_output=True,
        text=True,
    )
    value = result.stdout.strip()
    return value or None


def default_env_lookup(name: str) -> str | None:
    """Return a non-blank environment variable value when one is defined.

    Purpose:
        Provide a lightweight fallback lookup for author resolution
            without accepting blank environment values.

    Args:
        name (str): Environment variable name to read from the current process.

    Returns:
        str | None: The environment variable value when present and
            non-blank; otherwise `None`.

    Raises:
        None.

    Side Effects:
        Reads process environment state.
    """
    value = os.getenv(name)
    return value if value and value.strip() else None


def get_author(
    git_lookup: Callable[[str], str | None] = default_git_config_lookup,
    env_lookup: Callable[[str], str | None] = default_env_lookup,
) -> str:
    """Resolve the author name from Git configuration before falling back to USERNAME.

    Purpose:
        Produce the best available author label for the generated
            bug-entry markdown metadata.

    Args:
        git_lookup (Callable[[str], str | None]): Git-backed lookup
            function for configuration values.
        env_lookup (Callable[[str], str | None]): Environment-backed
            lookup function for fallback variables.

    Returns:
        str: The resolved author name, or `Unknown` when neither lookup yields a value.

    Raises:
        None.

    Side Effects:
        May read Git configuration and process environment state through
            the injected lookup callables.
    """
    author = git_lookup("user.name")
    if not author:
        author = env_lookup("USERNAME")
    if not author:
        return "Unknown"
    return author


def render_content(template: str, short_name: str, entry_date: str, author: str) -> str:
    """Apply bug-entry placeholder substitutions to the copied markdown template.

    Purpose:
        Replace the template placeholders with the concrete bug name,
            entry date, and author metadata.

    Args:
        template (str): Raw markdown template content copied from the
            selected source template.
        short_name (str): Validated short name used to replace the
            bug-name placeholder.
        entry_date (str): ISO date string inserted into the generated markdown.
        author (str): Author name inserted into the metadata block.

    Returns:
        str: Markdown content with all supported bug-entry placeholders replaced.

    Raises:
        None.

    Side Effects:
        None.
    """
    updated = template.replace("<bug-name>", short_name)
    updated = updated.replace("YYYY-MM-DD", entry_date)
    updated = updated.replace("- Author: name", f"- Author: {author}")
    return updated


class FileSystem(Protocol):
    def ensure_dir(self, path: Path) -> None: ...

    def copy_file(self, src: Path, dest: Path) -> None: ...

    def read_text(self, path: Path) -> str: ...

    def write_text(self, path: Path, content: str) -> None: ...


@dataclass
class RealFileSystem(FileSystem):
    def ensure_dir(self, path: Path) -> None:
        path.mkdir(parents=True, exist_ok=True)

    def copy_file(self, src: Path, dest: Path) -> None:
        shutil.copyfile(src, dest)

    def read_text(self, path: Path) -> str:
        return path.read_text(encoding="utf-8")

    def write_text(self, path: Path, content: str) -> None:
        path.write_text(content, encoding="utf-8")


def _is_insiders_session(
    env_lookup: Callable[[str], str | None] | None = None,
) -> bool:
    """Return whether the current process appears to be running inside VS Code Insiders.

    Purpose:
        Detect the current editor flavor so launcher selection can prefer
            `code-insiders` when the originating session is already Insiders.

    Args:
        env_lookup (Callable[[str], str | None]): Environment lookup used
            to read the known VS Code session signal variables.

    Returns:
        bool: `True` when any supported signal indicates an Insiders session;
            otherwise `False`.

    Raises:
        None.

    Side Effects:
        Reads process environment state through the injected lookup.
    """
    # Check the documented VS Code environment signals in a stable order so
    # the launcher behavior stays deterministic across bundled and root copies.
    lookup = env_lookup or default_env_lookup
    for variable_name in _INSIDERS_SIGNAL_NAMES:
        value = lookup(variable_name)
        if value and "insider" in value.lower():
            return True
    return False


def _resolve_code_cli(
    which_lookup: Callable[[str], str | None] | None = None,
    env_lookup: Callable[[str], str | None] | None = None,
) -> str | None:
    """Resolve the best VS Code CLI executable path for the current session.
        [code_cmd, "--reuse-window", *[file_path.as_posix() for file_path in files]],
    Purpose:
        Choose the correct CLI path while preserving fallback to the standard
            `code` command when an Insiders-specific executable is unavailable.

    Args:
        which_lookup (Callable[[str], str | None]): PATH lookup used to
            resolve CLI executables.
        env_lookup (Callable[[str], str | None]): Environment lookup used
            for Insiders-session detection.

    Returns:
        str | None: Resolved CLI executable path when one is available;
            otherwise `None`.

    Raises:
        None.

    Side Effects:
        Reads process environment state and PATH through the injected lookups.
    """
    resolved_which_lookup = which_lookup or shutil.which

    # Prefer the CLI that matches the current session first, then fall back to
    # the other supported executable name to preserve graceful behavior.
    candidate_names = (
        ("code-insiders", "code")
        if _is_insiders_session(env_lookup)
        else ("code", "code-insiders")
    )
    for candidate_name in candidate_names:
        resolved_command = resolved_which_lookup(candidate_name)
        if resolved_command:
            return resolved_command
    return None


def default_code_launcher(files: Iterable[Path]) -> bool:
    """Open created files in the matching VS Code session when possible.

    Purpose:
        Reuse the current VS Code or VS Code Insiders window for generated
            files while preserving the existing boolean success contract.

    Args:
        files (Iterable[Path]): Files to open in the editor session.

    Returns:
        bool: `True` when a supported VS Code CLI executable was resolved and
            invoked; otherwise `False`.

    Raises:
        None.

    Side Effects:
        Resolves editor CLI executables on PATH and invokes a subprocess when
            an appropriate command is available.
    """
    code_cmd = _resolve_code_cli()
    if not code_cmd:
        return False

    subprocess.run(  # noqa: S603
        [
            code_cmd,
            "--reuse-window",
            *[str(file_path).replace("\\", "/") for file_path in files],
        ],
        check=False,
    )
    return True


def create_bug_entry(
    short_name: str,
    workspace: Path | None = None,
    fs: FileSystem | None = None,
    author_provider: Callable[[], str] = get_author,
    code_launcher: Callable[[Iterable[Path]], bool] = default_code_launcher,
    entry_date: str | None = None,
    template_root: Path | None = None,
) -> Path:
    """Create a potential bug markdown file from the selected template root.

    Purpose:
        Materialize a new potential bug entry by copying a template,
            filling in metadata, and optionally opening the result in
            VS Code.

    Args:
        short_name (str): Validated kebab-case suffix for the generated
            bug-entry filename and content.
        workspace (Path | None): Workspace root that receives the
            generated file; defaults to the resolved repository or
            bundle workspace.
        fs (FileSystem | None): Filesystem adapter used for directory
            creation, file copy, and text I/O.
        author_provider (Callable[[], str]): Callable that resolves the
            author name inserted into the template.
        code_launcher (Callable[[Iterable[Path]], bool]): Callable that
            attempts to open the generated file in VS Code.
        entry_date (str | None): Optional ISO date override for
            deterministic generation.
        template_root (Path | None): Optional bundled feature-template
            root that overrides workspace-local templates.

    Returns:
        Path: Absolute path to the generated potential bug markdown file.

    Raises:
        ValueError: Raised when `short_name` violates the kebab-case contract.
        FileNotFoundError: Raised when the selected template cannot be
            copied from the resolved template path.

    Side Effects:
        Creates directories, copies and rewrites markdown files, and may
            invoke the VS Code launcher or print a warning.
    """
    validate_short_name(short_name)

    workspace_path = workspace or _resolve_workspace()
    filesystem = fs or RealFileSystem()
    date_str = entry_date or date.today().strftime("%Y-%m-%d")

    target_dir = workspace_path / "docs" / "features" / "potential"
    target = target_dir / f"{date_str}-{short_name}.md"
    if template_root is not None:
        template = template_root / "bug" / "potential_bug.md"
    else:
        template = (
            workspace_path
            / "docs"
            / "features"
            / "templates"
            / "bug"
            / "potential_bug.md"
        )

    filesystem.ensure_dir(target_dir)
    filesystem.copy_file(template, target)

    author = author_provider()
    content = filesystem.read_text(target)
    updated = render_content(content, short_name, date_str, author)
    filesystem.write_text(target, updated)

    if not code_launcher([target]):
        print("WARNING: VS Code 'code' command not found. Open file manually:")
        print(f"  {target}")

    return target


def parse_args() -> argparse.Namespace:
    """Parse CLI arguments for the potential bug entry workflow.

    Purpose:
        Define and read the command-line interface for creating a
            potential bug entry from a bundled or workspace template.

    Args:
        None.

    Returns:
        argparse.Namespace: Parsed CLI arguments containing the required
            short name and optional template root.

    Raises:
        SystemExit: Raised by `argparse` when required arguments are missing or invalid.

    Side Effects:
        Reads process command-line arguments and may emit argparse help or error text.
    """
    parser = argparse.ArgumentParser(
        description="Create a potential bug entry from the template."
    )
    parser.add_argument(
        "--short-name", required=True, help="Bug name in kebab-case (e.g., api-timeout)"
    )
    parser.add_argument(
        "--template-root",
        default=None,
        help="Bundled feature-templates dir (overrides workspace).",
    )
    return parser.parse_args()


def main() -> None:
    """Execute the CLI boundary for potential bug entry creation.

    Purpose:
        Bridge parsed CLI arguments into the bug-entry creation workflow
            and convert user-facing failures into exit codes.

    Args:
        None.

    Returns:
        None: The CLI exits by returning normally on success or raising
            `SystemExit` on handled failures.

    Raises:
        SystemExit: Raised with exit code 1 when validation or file lookup fails.

    Side Effects:
        Reads CLI arguments, may create bug-entry files, may launch VS
            Code, and prints user-facing error messages on handled
            failures.
    """
    args = parse_args()
    resolved_root = Path(args.template_root) if args.template_root else None
    try:
        create_bug_entry(short_name=args.short_name, template_root=resolved_root)
    except ValueError as exc:
        print(str(exc))
        raise SystemExit(1) from exc
    except FileNotFoundError as exc:
        print(f"Aborted: required file not found: {exc}")
        raise SystemExit(1) from exc


if __name__ == "__main__":
    main()

File: extensions/drm-copilot/src/lib/new-potential-bug-entry.ts
import * as fs from "node:fs";
import * as nodePath from "node:path";
import { type FileSystem, toPosixPath } from "./file-system";
import { type CommandRunner, SubprocessRunner } from "./subprocess-runner";

/**
 * In-process TypeScript port of the bundled Python
 * `resources/scripts/dev_tools/new_potential_bug_entry.py`.
 *
 * Purpose:
 *     Reproduce, verbatim, the observable behavior of the Python bug-entry
 *     creator: short-name validation, template selection/read, author
 *     resolution from git config / environment, placeholder rendering, target
 *     file write, and the optional VS Code launch. All external interactions
 *     (filesystem, git, PATH probing, environment, editor launch) flow through
 *     injectable seams so the module is hermetically testable and so the
 *     service/MCP path can pass a no-op editor launcher.
 *
 * Responsibilities:
 *     - {@link validateShortName}: enforce the kebab-case short-name contract
 *       with a byte-identical error message to the Python `ValueError`.
 *     - {@link renderContent}: replace-all placeholder substitution matching
 *       Python `str.replace` semantics.
 *     - Author resolution seams ({@link defaultGitConfigLookup},
 *       {@link defaultEnvLookup}, {@link getAuthor}).
 *     - Editor-launch seams ({@link isInsidersSession}, {@link resolveCodeCli},
 *       {@link defaultCodeLauncher}).
 *     - {@link createBugEntry}: orchestrate the full creation workflow.
 *
 * Invariants / Constraints:
 *     - The library entry point performs no direct `node:fs` /
 *       `node:child_process` access; those are confined to the injectable
 *       production defaults (env lookup, which lookup, git lookup, launcher),
 *       which callers can replace in tests.
 *     - No `copyFile` is added to the shared {@link FileSystem} interface; the
 *       Python copy-then-read collapses into a single template read followed by
 *       render-and-write, which is behaviorally identical for the observable
 *       outcome (target file content).
 */

/**
 * Anchored kebab-case pattern for bug-entry short names.
 *
 * Mirrors the Python `SHORT_NAME_PATTERN = re.compile(r"^[a-z0-9]+(-[a-z0-9]+)*$")`.
 * Python `fullmatch` against this anchored pattern is equivalent to
 * `RegExp.prototype.test` here because the pattern is anchored with `^...$`.
 */
export const SHORT_NAME_PATTERN = /^[a-z0-9]+(-[a-z0-9]+)*$/;

/**
 * VS Code session signal environment variable names, checked in a fixed order.
 *
 * Mirrors the Python `_INSIDERS_SIGNAL_NAMES` tuple. The order is preserved so
 * Insiders detection behaves identically to the Python source.
 */
const INSIDERS_SIGNAL_NAMES = [
  "TERM_PROGRAM_VERSION",
  "VSCODE_GIT_ASKPASS_MAIN",
  "TERM_PROGRAM",
  "VSCODE_IPC_HOOK_CLI",
] as const;

/**
 * Validate that a short name matches the repository kebab-case contract.
 *
 * Rejects invalid bug-entry short names before any filesystem or template work
 * begins, matching the Python `validate_short_name` contract.
 *
 * @param shortName Candidate short name that must use lowercase kebab-case
 *   tokens (letters/numbers, hyphen-separated).
 * @throws Error When `shortName` is blank or does not fully match
 *   {@link SHORT_NAME_PATTERN}. The message is byte-identical to the Python
 *   `ValueError` text.
 */
export function validateShortName(shortName: string): void {
  // A blank string or a value that does not fully match the anchored pattern is
  // rejected. The message text must remain byte-identical to the Python source.
  if (!shortName || !SHORT_NAME_PATTERN.test(shortName)) {
    throw new Error(
      `Aborted: '${shortName}' is invalid. Use kebab-case letters/numbers only ` +
        "(e.g., api-timeout).",
    );
  }
}

/**
 * Apply bug-entry placeholder substitutions to the markdown template content.
 *
 * Replaces every occurrence of each placeholder, matching Python `str.replace`
 * replace-all semantics. `String.prototype.replace` with a string argument
 * replaces only the first match in JavaScript, so this implementation uses
 * `split`/`join` to replace all occurrences.
 *
 * @param template Raw markdown template content.
 * @param shortName Validated short name replacing the `<bug-name>` placeholder.
 * @param entryDate ISO date string replacing the `YYYY-MM-DD` placeholder.
 * @param author Author name replacing the `- Author: name` metadata line.
 * @returns Markdown content with all supported placeholders replaced.
 */
export function renderContent(
  template: string,
  shortName: string,
  entryDate: string,
  author: string,
): string {
  // Replace-all every placeholder via split/join so all occurrences are
  // substituted, matching Python `str.replace` (which replaces all matches).
  let updated = template.split("<bug-name>").join(shortName);
  updated = updated.split("YYYY-MM-DD").join(entryDate);
  updated = updated.split("- Author: name").join(`- Author: ${author}`);
  return updated;
}

/**
 * Resolve a git configuration value through the injected command runner.
 *
 * Production default for the author git lookup. Runs `git config <key>` with
 * `allowError: true` so a missing git executable or a non-zero exit yields
 * `undefined` rather than throwing, mirroring the Python `default_git_config_
 * lookup` behavior (which probes `shutil.which("git")` and returns `None` on
 * absence). The injected {@link CommandRunner} surfaces a spawn failure as a
 * non-zero `code`, which this function treats as "no value".
 *
 * @param runner Command runner used to invoke git.
 * @param key Git configuration key to read, such as `user.name`.
 * @returns The trimmed configuration value when non-empty; otherwise
 *   `undefined`.
 */
export function defaultGitConfigLookup(
  runner: CommandRunner,
  key: string,
): string | undefined {
  const result = runner.run(["git", "config", key], { allowError: true });
  // A non-zero exit (including a failed spawn when git is absent) means no
  // value is available, mirroring the Python `which("git")` guard returning None.
  if (result.code !== 0) {
    return undefined;
  }
  const value = result.stdout.trim();
  return value.length > 0 ? value : undefined;
}

/**
 * Return a non-blank environment variable value when one is defined.
 *
 * Production default for the author/insiders environment lookup. Mirrors the
 * Python `default_env_lookup`: reads `process.env[name]` and returns it only
 * when present and non-blank after trimming, else `undefined`.
 *
 * @param name Environment variable name to read.
 * @returns The value when present and non-blank; otherwise `undefined`.
 */
export function defaultEnvLookup(name: string): string | undefined {
  const value = process.env[name];
  return value !== undefined && value.trim().length > 0 ? value : undefined;
}

/**
 * Resolve the author name, preferring git config before the USERNAME fallback.
 *
 * Mirrors the Python `get_author`: query `user.name` via the git lookup first;
 * if blank/undefined, read the `USERNAME` environment variable; if still
 * blank/undefined, return the literal `"Unknown"`.
 *
 * @param gitLookup Git-backed lookup for configuration values.
 * @param envLookup Environment-backed lookup for fallback variables.
 * @returns The resolved author name, or `"Unknown"` when neither yields a value.
 */
export function getAuthor(
  gitLookup: (key: string) => string | undefined,
  envLookup: (name: string) => string | undefined,
): string {
  // Prefer the repository's configured author; fall back to USERNAME; then to
  // the literal "Unknown". Ordering matches the Python source exactly.
  let author = gitLookup("user.name");
  if (!author) {
    author = envLookup("USERNAME");
  }
  if (!author) {
    return "Unknown";
  }
  return author;
}

/**
 * Determine whether the current process appears to run inside VS Code Insiders.
 *
 * Mirrors the Python `_is_insiders_session`: probe the documented VS Code
 * session signal variables in a fixed order; a value containing the substring
 * `insider` (case-insensitive) indicates an Insiders session.
 *
 * @param envLookup Environment lookup used to read the signal variables.
 * @returns `true` when any supported signal indicates Insiders; else `false`.
 */
export function isInsidersSession(
  envLookup: (name: string) => string | undefined,
): boolean {
  // Check the documented VS Code environment signals in a stable order so the
  // launcher behavior stays deterministic, matching the Python ordering.
  for (const variableName of INSIDERS_SIGNAL_NAMES) {
    const value = envLookup(variableName);
    if (value !== undefined && value.toLowerCase().includes("insider")) {
      return true;
    }
  }
  return false;
}

/**
 * Resolve the best VS Code CLI executable for the current session.
 *
 * Mirrors the Python `_resolve_code_cli`: when the session signals Insiders,
 * prefer `code-insiders` then fall back to `code`; otherwise prefer `code`
 * then fall back to `code-insiders`. Returns the first resolved executable.
 *
 * @param whichLookup PATH lookup resolving a CLI name to an executable path.
 * @param envLookup Environment lookup used for Insiders-session detection.
 * @returns The resolved CLI executable path, or `undefined` when none resolves.
 */
export function resolveCodeCli(
  whichLookup: (name: string) => string | undefined,
  envLookup: (name: string) => string | undefined,
): string | undefined {
  // Prefer the CLI matching the current session first, then fall back to the
  // other supported executable name to preserve graceful behavior.
  const candidateNames = isInsidersSession(envLookup)
    ? (["code-insiders", "code"] as const)
    : (["code", "code-insiders"] as const);
  for (const candidateName of candidateNames) {
    const resolvedCommand = whichLookup(candidateName);
    if (resolvedCommand) {
      return resolvedCommand;
    }
  }
  return undefined;
}

/**
 * Probe the current process PATH for an executable and return its path.
 *
 * Production default for the {@link resolveCodeCli} `whichLookup` seam. Mirrors
 * `shutil.which` using the same PATH/PATHEXT probing approach as
 * `command-runtime.ts`. Injectable so tests never touch the real PATH.
 *
 * @param executable The executable name to probe, without a file extension.
 * @returns The resolved absolute path when found; otherwise `undefined`.
 */
export function defaultWhichLookup(executable: string): string | undefined {
  const pathValue = process.env["PATH"] ?? "";
  const pathParts = pathValue
    .split(nodePath.delimiter)
    .filter((part) => part.length > 0);
  const pathExtensions =
    process.platform === "win32"
      ? (process.env["PATHEXT"] ?? ".COM;.EXE;.BAT;.CMD")
          .split(";")
          .filter((part) => part.length > 0)
      : [""];

  // Probe each PATH directory against each allowed extension so resolution
  // behaves consistently across Windows and non-Windows environments.
  for (const directory of pathParts) {
    for (const extension of pathExtensions) {
      const candidate = nodePath.join(
        directory,
        process.platform === "win32" ? `${executable}${extension}` : executable,
      );
      if (fs.existsSync(candidate)) {
        return candidate;
      }
    }
  }
  return undefined;
}

/** Injectable dependencies for {@link defaultCodeLauncher}. */
export interface CodeLauncherDeps {
  /** Command runner used to invoke the resolved CLI executable. */
  readonly runner: CommandRunner;
  /** PATH lookup resolving a CLI name to an executable path. */
  readonly whichLookup: (name: string) => string | undefined;
  /** Environment lookup used for Insiders-session detection. */
  readonly envLookup: (name: string) => string | undefined;
}

/**
 * Open the created files in the matching VS Code session when possible.
 *
 * Production default editor launcher. Mirrors the Python `default_code_
 * launcher`: resolve the CLI via {@link resolveCodeCli}; if none resolves,
 * return `false`; otherwise invoke `<cli> --reuse-window <files...>` (with
 * forward-slash file paths) through the injected runner and return `true`.
 *
 * @param files Files to open in the editor session.
 * @param deps Injected runner / which / env seams.
 * @returns `true` when a CLI was resolved and invoked; otherwise `false`.
 */
export function defaultCodeLauncher(
  files: readonly string[],
  deps: CodeLauncherDeps,
): boolean {
  const codeCmd = resolveCodeCli(deps.whichLookup, deps.envLookup);
  if (!codeCmd) {
    return false;
  }

  // Invoke the resolved CLI with the same argument shape as the Python source:
  // `--reuse-window` followed by forward-slash-normalized file paths. The run
  // tolerates a non-zero exit (allowError) to preserve the boolean success
  // contract once a CLI is resolved.
  deps.runner.run(
    [
      codeCmd,
      "--reuse-window",
      ...files.map((filePath) => filePath.replace(/\\/g, "/")),
    ],
    { allowError: true },
  );
  return true;
}

/** Options for {@link createBugEntry}. */
export interface CreateBugEntryOptions {
  /** Validated kebab-case short name for the generated filename and content. */
  readonly shortName: string;
  /** Workspace root that receives the generated file. */
  readonly workspace: string;
  /** Injected filesystem for directory creation and text I/O. */
  readonly fs: FileSystem;
  /** Author resolver; defaults to {@link getAuthor} with production lookups. */
  readonly authorProvider?: () => string;
  /** Editor launcher; defaults to {@link defaultCodeLauncher}. */
  readonly codeLauncher?: (files: readonly string[]) => boolean;
  /** Optional ISO date override (`YYYY-MM-DD`) for deterministic generation. */
  readonly entryDate?: string;
  /** Optional bundled feature-template root overriding workspace templates. */
  readonly templateRoot?: string;
  /** Optional log sink for the editor-not-found warning lines. */
  readonly log?: (message: string) => void;
}

/**
 * Build the production author provider backed by git config and environment.
 *
 * @param runner Command runner used by the default git lookup.
 * @returns A zero-argument author resolver matching the Python default chain.
 */
function buildDefaultAuthorProvider(runner: CommandRunner): () => string {
  return () =>
    getAuthor((key) => defaultGitConfigLookup(runner, key), defaultEnvLookup);
}

/**
 * Resolve today's date as an ISO `YYYY-MM-DD` string.
 *
 * Mirrors the Python `date.today().strftime("%Y-%m-%d")`. Used only when no
 * explicit `entryDate` override is supplied.
 *
 * @returns The current local date formatted as `YYYY-MM-DD`.
 */
function todayIsoDate(): string {
  const now = new Date();
  const year = now.getFullYear().toString().padStart(4, "0");
  const month = (now.getMonth() + 1).toString().padStart(2, "0");
  const day = now.getDate().toString().padStart(2, "0");
  return `${year}-${month}-${day}`;
}

/**
 * Create a potential bug markdown file from the selected template root.
 *
 * Materializes a new potential bug entry by reading a template, filling in
 * metadata, writing the rendered result, and optionally opening it in VS Code.
 * The Python copy-then-read-then-write is collapsed into a single template read
 * followed by render-and-write, which is behaviorally identical for the target
 * file content (the only observable filesystem outcome).
 *
 * Flow:
 *     1. Validate the short name (before any filesystem work).
 *     2. Resolve the date string, target directory, target path, and template
 *        path (templateRoot override vs workspace fallback).
 *     3. Ensure the target directory exists.
 *     4. Read the template; resolve the author; render placeholders; write the
 *        target file.
 *     5. Attempt the editor launch; on failure, emit the two warning lines.
 *
 * @param options See {@link CreateBugEntryOptions}.
 * @returns The created absolute target path (forward-slash joined per the
 *   injected filesystem and `node:path` conventions).
 * @throws Error When the short name is invalid (before any write) or when the
 *   template path is absent (surfaced as a file-not-found read error).
 */
export function createBugEntry(options: CreateBugEntryOptions): string {
  validateShortName(options.shortName);

  const dateStr = options.entryDate ?? todayIsoDate();
  // The service/MCP path always injects both seams. The defaults below back
  // standalone/CLI-equivalent callers and construct a real SubprocessRunner.
  const authorProvider =
    options.authorProvider ??
    buildDefaultAuthorProvider(new SubprocessRunner());
  const codeLauncher =
    options.codeLauncher ??
    ((files: readonly string[]) =>
      defaultCodeLauncher(files, {
        runner: new SubprocessRunner(),
        whichLookup: defaultWhichLookup,
        envLookup: defaultEnvLookup,
      }));

  // Paths are joined with forward-slash separators (`node:path` POSIX join over
  // toPosixPath-normalized inputs) to match the F1 FileSystem convention and the
  // Python `Path` POSIX-style outputs, so the returned/written path is stable
  // and OS-neutral.
  const targetDir = nodePath.posix.join(
    toPosixPath(options.workspace),
    "docs",
    "features",
    "potential",
  );
  const targetPath = nodePath.posix.join(
    targetDir,
    `${dateStr}-${options.shortName}.md`,
  );

  // Branch decision: a provided templateRoot overrides the workspace-local
  // template directory, matching the Python `template_root is not None` branch
  // (the service path always passes templateRoot; the fallback mirrors the CLI).
  const templatePath =
    options.templateRoot !== undefined
      ? nodePath.posix.join(
          toPosixPath(options.templateRoot),
          "bug",
          "potential_bug.md",
        )
      : nodePath.posix.join(
          toPosixPath(options.workspace),
          "docs",
          "features",
          "templates",
          "bug",
          "potential_bug.md",
        );

  options.fs.ensureDir(targetDir);
  const template = options.fs.readTextFile(templatePath);
  const author = authorProvider();
  const updated = renderContent(template, options.shortName, dateStr, author);
  options.fs.writeTextFile(targetPath, updated);

  // When the editor launch is unavailable (or a no-op in the MCP/service path),
  // emit the manual-open guidance lines, preserving the Python warning text.
  if (!codeLauncher([targetPath])) {
    options.log?.(
      "WARNING: VS Code 'code' command not found. Open file manually:",
    );
    options.log?.(`  ${targetPath}`);
  }

  return targetPath;
}
